src/OrchardCore.Modules/OrchardCore.Taxonomies/Fields/TagNamesExtensions.cs

```diff
@@ -0,0 +1,29 @@
+using System;
+using Newtonsoft.Json.Linq;
+
+namespace OrchardCore.Taxonomies.Fields
+{
+    public static class TagNamesExtensions
+    {
+        /// <summary>
+        /// Tag names are a less well known property of a taxonomy field
+        /// managed by the tags editor and tags display mode. 
+        /// </summary>
+        public static string[] GetTagNames(this TaxonomyField taxonomyField)
+        {
+            var tagNames = taxonomyField.Content["TagNames"] as JArray;
+
+            return tagNames != null ? tagNames.ToObject<string[]>() : Array.Empty<string>();
+        }
+
+        /// <summary>
+        /// Tags names are a less well known property of a taxonomy field
+        /// managed by the tags editor and tags display mode.
+        /// They are updated only when saving the content item.
+        /// </summary>
+        public static void SetTagNames(this TaxonomyField taxonomyField, string[] tagNames)
+        {
+            taxonomyField.Content["TagNames"] = JArray.FromObject(tagNames);
+        }
```

Comment:
@sebastienros implementing the `TagNames` `ContentElement` as an `IEnumerable` does not work because the elements dictionary expects `JObject` and won't take a `JArray`

eg https://github.com/OrchardCMS/OrchardCore/blob/a5e11d392ac191b6b9d1baa99a1753fb1fdbf4fa/src/OrchardCore/OrchardCore.ContentManagement.Abstractions/ContentExtensions.cs#L154

So implemented instead directly against the `Content` property

Which makes the accessor in liquid `Model.ContentItem....TaxonomyFieldName.TagNames`